inspirehep/submissions/marshmallow/literature.py

```diff
@@ -0,0 +1,217 @@
+                thesis_info, "institutions[0].name", default=missing
+            ),
+            "supervisors": self.get_authors_by_role(data, "supervisor"),
+        }
+
+    @staticmethod
+    def get_authors_by_role(data, role):
+        authors = data.get("authors", [])
+        return [
+            Literature.record_author_to_form_author(author)
+            for author in authors
+            if role in author.get("inspire_roles", ["author"])
+        ] or missing
+
+    @staticmethod
+    def record_author_to_form_author(author):
+        form_author = {"full_name": author["full_name"]}
+        affiliation = get_value(author, "affiliations[0].value")
+        if affiliation is not None:
+            form_author["affiliation"] = affiliation
+        return form_author
+
+    @staticmethod
+    def get_publication_artid_or_page_range_or_missing(publication_info):
+        artid = publication_info.get("artid")
+        if artid is not None:
+            return artid
+        start_page = publication_info.get("page_start")
+        end_page = publication_info.get("page_end")
+
+        if start_page is not None and end_page is not None:
+            return "{}{}{}".format(start_page, PAGE_RANGE_SEPARATOR, end_page)
+        return missing
+
+    @post_load
+    def build_literature(self, data):
+        literature = LiteratureBuilder()
+
+        literature.add_document_type(data["document_type"])
+
+        literature.add_arxiv_eprint(data.get("arxiv_id"))
+        literature.add_doi(data.get("doi"))
+
+        # TODO: pdf_link has to be passed somehow to the workflow unfortunately
+        literature.add_url(data.get("additional_link"))
+
+        literature.add_title(data["title"], source="submitter")
+        literature.add_language(data.get("language"))
+        literature.add_inspire_categories(data.get("subjects"))
+
+        for author in data.get("authors", []):
+            record_author = literature.make_author(
+                author.get("full_name"), affiliations=[author.get("affiliation")]
+            )
+            literature.add_author(record_author)
+
+        literature.add_collaboration(data.get("collaboration"))
+        literature.add_accelerator_experiments_legacy_name(data.get("experiment"))
+        # TODO: source=submitter?
+        literature.add_abstract(data.get("abstract"), source="submitter")
+
+        for report_number in data.get("report_numbers", []):
+            literature.add_report_number(report_number, source="submitter")
+
+        parsed_page_range_or_artid = self.parse_page_range_or_artid(
+            data.get("page_range")
+        )
+        literature.add_publication_info(
+            journal_title=data.get("journal_title"),
+            journal_volume=data.get("volume"),
+            journal_issue=data.get("issue"),
+            artid=parsed_page_range_or_artid.get("artid"),
+            page_start=data.get("start_page")
+            or parsed_page_range_or_artid.get("page_start"),
```

Comment:
this is kinda weird, because if it's an `article`, `page_range` is provided and if it's a `chapter` then `page_start` and `page_end` are provided. But I will try to write this part more obviously